File: backend-rest/src/main/java/com/tp_distribuidos/backend_rest/models/entities/Work.java
package com.tp_distribuidos.backend_rest.models.entities;

import com.tp_distribuidos.backend_rest.enums.WorkAvailability;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "works", indexes = {
        @Index(name = "idx_works_artist_id", columnList = "artist_id"),
        @Index(name = "idx_works_title", columnList = "title"),
        @Index(name = "idx_works_era", columnList = "era")
})
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Work {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 255)
    private String title;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "artist_id", nullable = false)
    private Artist artist;

    @Column(name = "image_url", nullable = false, length = 500)
    private String imageUrl;

    @Column(name = "creation_year")
    private Integer creationYear;

    @Column(length = 100)
    private String technique;

    @Column(length = 100)
    private String dimensions;

    @Column(length = 100)
    private String era;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(length = 255)
    private String location;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private WorkAvailability availability;

    @OneToMany(mappedBy = "work", fetch = FetchType.LAZY)
    private List<Comment> comments = new ArrayList<>();

    public static Work of(String title, Artist artist, String imageUrl, Integer creationYear,
                          String technique, String dimensions, String era, String description,
                          String location, WorkAvailability availability) {
        Work work = new Work();
        work.title = title;
        work.artist = artist;
        work.imageUrl = imageUrl;
        work.creationYear = creationYear;
        work.technique = technique;
        work.dimensions = dimensions;
        work.era = era;
        work.description = description;
        work.location = location;
        work.availability = availability;
        return work;
    }
}

File: backend-graphql/work.go
package main

import "github.com/graphql-go/graphql"

type Artist struct {
	Name      string `json:"name"`
	Biography string `json:"biography"`
}

type Comment struct {
	User string `json:"user"`
	Text string `json:"text"`
	Date string `json:"date"`
}

type Work struct {
	ID           string    `json:"id"`
	Title        string    `json:"title"`
	Artist       Artist    `json:"artist"`
	ImageURL     string    `json:"image_url"`
	CreationYear int       `json:"creation_year"`
	Technique    string    `json:"technique"`
	Dimensions   string    `json:"dimensions"`
	Era          string    `json:"era"`
	Description  string    `json:"description"`
	Location     string    `json:"location"`
	Availability bool      `json:"availability"`
	Comments     []Comment `json:"comments"`
}

var artistType = graphql.NewObject(graphql.ObjectConfig{
	Name: "Artist",
	Fields: graphql.Fields{
		"name":      &graphql.Field{Type: graphql.String},
		"biography": &graphql.Field{Type: graphql.String},
	},
})

var commentType = graphql.NewObject(graphql.ObjectConfig{
	Name: "Comment",
	Fields: graphql.Fields{
		"user": &graphql.Field{Type: graphql.String},
		"text": &graphql.Field{Type: graphql.String},
		"date": &graphql.Field{Type: graphql.String},
	},
})

var workType = graphql.NewObject(graphql.ObjectConfig{
	Name: "Work",
	Fields: graphql.Fields{
		"id":            &graphql.Field{Type: graphql.ID},
		"title":         &graphql.Field{Type: graphql.String},
		"artist":        &graphql.Field{Type: artistType},
		"image_url":     &graphql.Field{Type: graphql.String},
		"creation_year": &graphql.Field{Type: graphql.Int},
		"technique":     &graphql.Field{Type: graphql.String},
		"dimensions":    &graphql.Field{Type: graphql.String},
		"era":           &graphql.Field{Type: graphql.String},
		"description":   &graphql.Field{Type: graphql.String},
		"location":      &graphql.Field{Type: graphql.String},
		"availability":  &graphql.Field{Type: graphql.Boolean},
		"comments":      &graphql.Field{Type: graphql.NewList(commentType)},
	},
})
